File: leetcode/algorithms/144_binary_tree_preorder_traversal/main.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * Iteration
 *
 * Complexities:
 *   N - Number of nodes in `root`
 *   H - Height of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(H)
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function (root) {
  const result = [];
  const totalNodes = [];

  if (root) {
    totalNodes.push(root);

    while (totalNodes.length > 0) {
      const node = totalNodes.pop();
      result.push(node.val);

      if (node.right) {
        totalNodes.push(node.right);
      }

      if (node.left) {
        totalNodes.push(node.left);
      }
    }
  }

  return result;
};


// Solution
/**
 * Solution 1
 *
 * Recursion
 *
 * Complexities:
 *   N - Number of nodes in `root`
 *   H - Height of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(H)
 */
var solution1 = function (root) {
  const result = [];

  function traverse(node) {
    if (!node) {
      return;
    }

    result.push(node.val);
    traverse(node.left);
    traverse(node.right);
  }

  traverse(root);
  return result;
};

/**
 * Solution 2
 *
 * Iteration
 *
 * Complexities:
 *   N - Number of nodes in `root`
 *   H - Height of `root`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(H)
 */
var solution2 = function (root) {
  if (!root) {
    return [];
  }

  const result = [];
  const stack = [root];

  while (stack.length > 0) {
    const node = stack.pop();
    result.push(node.val);

    if (node.right !== null) {
      stack.push(node.right);
    }

    if (node.left !== null) {
      stack.push(node.left);
    }
  }

  return result;
};

File: leetcode/algorithms/144_binary_tree_preorder_traversal/main.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BinaryTreePreorderTraversal:
    """
    # Iteration
    #
    # Complexities:
    #   N - Number of nodes in `root`
    #   H - Height of `root`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(H)
    """
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        totalNodes = []

        if root is not None:
            totalNodes.append(root)

            while len(totalNodes) > 0:
                node = totalNodes.pop()
                result.append(node.val)

                if node.right is not None:
                    totalNodes.append(node.right)

                if node.left is not None:
                    totalNodes.append(node.left)

        return result


    # Solution
    """
    # Solution 1
    #
    # Recursion
    #
    # Complexities:
    #   N - Number of nodes in `root`
    #   H - Height of `root`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(H)
    """
    def solution1(self, root: Optional[TreeNode]) -> List[int]:
        result = []

        def traverse(node):
            if not node:
                return

            result.append(node.val)

            traverse(node.left)

            traverse(node.right)

        traverse(root)

        return result

    """
    # Solution 2
    #
    # Iteration
    #
    # Complexities:
    #   N - Number of nodes in `root`
    #   H - Height of `root`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(H)
    """
    def solution2(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        result = []
        stack = [root]

        while stack:
            node = stack.pop()

            result.append(node.val)

            if node.right:
                stack.append(node.right)

            if node.left:
                stack.append(node.left)

        return result
